pkg/resources/grant_application_role.go

```diff
@@ -0,0 +1,199 @@
+package resources
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"strings"
+
+	"github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/internal/provider"
+
+	"github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/helpers"
+	"github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/sdk"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+var grantApplicationRoleSchema = map[string]*schema.Schema{
+	"name": {
+		Type:             schema.TypeString,
+		Required:         true,
+		Description:      "Specifies the identifier for the application role to grant.",
+		ForceNew:         true,
+		ValidateDiagFunc: IsValidIdentifier[sdk.DatabaseObjectIdentifier](),
+	},
+	"parent_role_name": {
+		Type:             schema.TypeString,
+		Optional:         true,
+		Description:      "The fully qualified name of the account role on which application role will be granted.",
+		ForceNew:         true,
+		ValidateDiagFunc: IsValidIdentifier[sdk.AccountObjectIdentifier](),
+		ExactlyOneOf: []string{
+			"parent_role_name",
+			"application_name",
+		},
+	},
+	"application_name": {
+		Type:             schema.TypeString,
+		Optional:         true,
+		Description:      "The fully qualified name of the application on which application role will be granted.",
+		ForceNew:         true,
+		ValidateDiagFunc: IsValidIdentifier[sdk.AccountObjectIdentifier](),
+		ExactlyOneOf: []string{
+			"parent_role_name",
+			"application_name",
+		},
+	},
+}
+
+func GrantApplicationRole() *schema.Resource {
+	return &schema.Resource{
+		Create: CreateGrantApplicationRole,
+		Read:   ReadGrantApplicationRole,
+		Delete: DeleteGrantApplicationRole,
+		Schema: grantApplicationRoleSchema,
+		Importer: &schema.ResourceImporter{
+			StateContext: func(ctx context.Context, d *schema.ResourceData, m interface{}) ([]*schema.ResourceData, error) {
+				parts := strings.Split(d.Id(), helpers.IDDelimiter)
+				if len(parts) != 3 {
+					return nil, fmt.Errorf("invalid ID specified: %v, expected <name>|<object_type>|<target_identifier>", d.Id())
+				}
+				if err := d.Set("name", parts[0]); err != nil {
+					return nil, err
+				}
+				switch parts[1] {
+				case "ROLE":
+					if err := d.Set("parent_role_name", strings.Trim(parts[2], "\"")); err != nil {
+						return nil, err
+					}
+				case "APPLICATION":
+					if err := d.Set("application_name", strings.Trim(parts[2], "\"")); err != nil {
```

Comment:
I'm not sure if calling `strings.Trim` is right, when e.g. in the Create we're saving the id with `FullyQualifiedName`. That makes it inconsistent with what is saved to the state. There were some issues created about it, so maybe we should just always save it with quotes?

Reply:
Trim removed.